File: rules/pii/sensitive-log-field.py
import logging

logger = logging.getLogger(__name__)


def login_bad(user, password):
    # ruleid: tqa-pii-sensitive-log-field-py
    logger.info("login attempt", extra={"user": user.id, "password": password})

    # ruleid: tqa-pii-sensitive-log-field-py
    logger.error("auth failed", access_token=user.token)


def login_good(user, password):
    # ok: tqa-pii-sensitive-log-field-py
    logger.info("login attempt", extra={"user_id": user.id, "password_hash": hash_pw(password)})

    # ok: tqa-pii-sensitive-log-field-py
    logger.error("auth failed", token_type="bearer", auth_method="oauth")

    # ok: tqa-pii-sensitive-log-field-py
    logger.info("payment", extra={"card_last4": card.last4})

    # ok: tqa-pii-sensitive-log-field-py
    logger.info("done", extra={"duration_ms": 12})

File: rules/pii/sensitive-log-field.go
package handler

import (
	"log/slog"

	"go.uber.org/zap"
)

func loginBad(email, password string, logger *zap.Logger) {
	// ruleid: tqa-pii-sensitive-log-field-go
	slog.Info("login attempt", slog.String("password", password))

	// ruleid: tqa-pii-sensitive-log-field-go
	logger.Info("auth", zap.Any("access_token", getToken()))
}

func loginGood(email, password string, logger *zap.Logger) {
	// ok: tqa-pii-sensitive-log-field-go
	slog.Info("login attempt", slog.String("password_hash", hash(password)))

	// ok: tqa-pii-sensitive-log-field-go
	logger.Info("auth", zap.String("token_type", "bearer"))

	// ok: tqa-pii-sensitive-log-field-go
	logger.Info("payment", zap.String("card_last4", last4(card)))

	// ok: tqa-pii-sensitive-log-field-go
	slog.Info("login ok", slog.String("account_id", userID))
}